### PR TITLE
fix(ZNTA-1083): add offline PO upload for File projects

### PR DESCRIPTION
**Pull request by djansen-redhat**
JIRA issue URL: https://zanata.atlassian.net/browse/ZNTA-1083

Allow the upload of offlinepo to a File project document. Give sufficient warning and documentation assistance for the use of raw files.

## Checklist

- JIRA link
- Check target branch
- Make sure all commit statuses are green or otherwise documented reasons to ignore
- QA needs to evaluate against the JIRA ticket
- Changed files and commits make sense for this PR

See [Zanata Development Guidelines](https://github.com/zanata/zanata-platform/wiki/Development-Guidelines) more for information.

----
*This template can be updated in .github/PULL_REQUEST_TEMPLATE.md*

### CODE REVIEW

File: docs/user-guide/documents/upload-documents.md [hunk 1]
```diff
@@ -52,6 +52,16 @@ You will need to [Create a project](/user-guide/projects/create-project) and [Cr
     - `Merge` option - If unchecked, uploaded translations overrides current translation, otherwise, it will merge with current translation in system.
     - `My translations` option - Indicates if all uploaded translations were translated by you.
 
+#### Translation document formats
+Translation files, typically of the Gettext .po format, can be uploaded for a document in the selected language.
+There is a limited set of document formats that support upload of a translated raw document. These are:
+ * Properties (.properties)
```
Comment (seanf) [inline docs/user-guide/documents/upload-documents.md]: Does this really work though? You can't currently upload .properties to the FileService, nor to the web UI, AFAIK.

File: server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java [hunk 2]
```diff
@@ -20,31 +20,153 @@
...
+        String poContent = "msgid \"\"\n" +
+                "msgstr \"\"\n" +
+                "\"Project-Id-Version: test-master\\n\"\n" +
+                "\n" +
+                "#, c-format\n" +
+                "msgid \"Thing 1\"\n" +
+                "msgid_plural \"%d files removed\"\n" +
+                "msgstr[0] \"1 aoeuaouaou\"\n" +
+                "msgstr[1] \"%d aoeuaouao\"\n" +
+                "msgstr[2] \"\"";
```
Comment (seanf) [inline server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java]: This test would be easier to read in Kotlin as a multi-line string.

File: server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java [hunk 3]
```diff
@@ -20,31 +20,153 @@
...
+    @Test
+    @InRequestScope
+    public void parseAdapterFileInFileProject() throws Exception {
+        String project = "test";
+        String version = "master";
+        String docId = "test.properties";
+        String plaintextContent = "first: test message";
+        File tempFile = File.createTempFile("test", ".properties");
+        Files.write(tempFile.toPath(), plaintextContent.getBytes());
+        InputStream stream = new FileInputStream(tempFile);
```
Comment (seanf) [inline server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java]: You forgot to close it. But a ByteArrayInputStream would avoid the need for a temp file.

File: server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java [hunk 4]
```diff
@@ -20,31 +20,153 @@
...
+
+    @Test
+    @InRequestScope
+    public void parseAdapterFileInFileProject() throws Exception {
+        String project = "test";
+        String version = "master";
+        String docId = "test.properties";
+        String plaintextContent = "first: test message";
+        File tempFile = File.createTempFile("test", ".properties");
+        Files.write(tempFile.toPath(), plaintextContent.getBytes());
```
Comment (seanf) [inline server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java]: use getBytes(java.nio.charset.StandardCharsets#UTF_8)

Comment (djansen-redhat) [reply server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java]: Done

File: server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java (hunk 2 above)
Comment (djansen-redhat) [reply server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java]: /runs

File: server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java (hunk 3 above)
Comment (djansen-redhat) [reply server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java]: Done

File: docs/user-guide/documents/upload-documents.md (hunk 1 above)
Comment (djansen-redhat) [reply docs/user-guide/documents/upload-documents.md]: Done (removed)

**Comment by djansen-redhat** [comment]
Reader note: this change is a precursor to https://github.com/zanata/zanata-platform/pull/849

**Comment by seanf** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java, line 125 at r1](https://reviewable.io/reviews/zanata/zanata-platform/858#-LDjfeHGMd1XXIsnWPwY-r1-125:-LDo9Q-0AuXgSyqfZZJD:b-og7fdn) ([raw file](https://github.com/zanata/zanata-platform/blob/4c3b9a357aece92d15c408fc4fac3eabc8c10c74/server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java#L125)):*
<details><summary><i>Previously, djansen-redhat (Damian Jansen) wrote…</i></summary><blockquote>

Done
</blockquote></details>

That still looks like an unclosed FileInputStream.

Try something like this:

    InputStream stream = new ByteArrayInputStream(plaintextContent.getBytes(StandardCharsets.UTF_8));

---

*[server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java, line 96 at r2](https://reviewable.io/reviews/zanata/zanata-platform/858#-LDo9ilr7FrC0cTALmbm:-LDo9ilr7FrC0cTALmbn:b-r4ckgl) ([raw file](https://github.com/zanata/zanata-platform/blob/0f275ea9d07d3efee20dd45859ffb9ced8cafd9d/server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java#L96)):*
> ```Java
>         File tempFile = File.createTempFile("test", ".po");
>         Files.write(tempFile.toPath(), poContent.getBytes());
>         InputStream stream = new FileInputStream(tempFile);
> ```

try something like this:

    InputStream stream = new ByteArrayInputStream(poContent.getBytes(StandardCharsets.UTF_8));

---

*[server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java, line 114 at r2](https://reviewable.io/reviews/zanata/zanata-platform/858#-LDoC2rk2tPrJaPlc-vM:-LDoC2rlE17XUDLciiTa:b95t89g) ([raw file](https://github.com/zanata/zanata-platform/blob/0f275ea9d07d3efee20dd45859ffb9ced8cafd9d/server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java#L114)):*
> ```Java
>                 "ru", project, version, docId, Optional.absent());
>         assertThat(translationsResource.getTextFlowTargets().get(0)
>                 .getContents().get(0)).isEqualTo("1 aoeuaouaou");
> ```

Just to make this test a bit better, could you also check the resId? I think it's hash(msgid+msgctxt).

---

*[server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java, line 149 at r2](https://reviewable.io/reviews/zanata/zanata-platform/858#-LDoCBjOC3zGgWnXqvV0:-LDoCBjOC3zGgWnXqvV1:b-40a2v0) ([raw file](https://github.com/zanata/zanata-platform/blob/0f275ea9d07d3efee20dd45859ffb9ced8cafd9d/server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java#L149)):*
> ```Java
>         stream.close();
>         assertThat(translationsResource.getTextFlowTargets().get(0)
>                 .getContents().get(0)).isEqualTo("test message");
> ```

Just to make this test a bit better, could you also check the resId? I think it should be 'first', since that's the key in the Properties file.

---

*[server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java, line 159 at r2](https://reviewable.io/reviews/zanata/zanata-platform/858#-LDoAAWo4TCyFnAbil6K:-LDoAAWo4TCyFnAbil6L:b-3evbhn) ([raw file](https://github.com/zanata/zanata-platform/blob/0f275ea9d07d3efee20dd45859ffb9ced8cafd9d/server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java#L159)):*
> ```Java
>         String docId = "test.pot";
>         File tempFile = File.createTempFile("test", ".ts");
>         InputStream stream = new FileInputStream(tempFile);
> ```

You could use ByteArrayInputStream with `new byte[0]`, or org.apache.commons.io.input.NullInputStream.

---


*Comments from [Reviewable](https://reviewable.io/reviews/zanata/zanata-platform/858)*

**Comment by djansen-redhat** [comment]
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java, line 125 at r1](https://reviewable.io/reviews/zanata/zanata-platform/858#-LDjfeHGMd1XXIsnWPwY-r1-125:-LDoJ7Vy88HreSxEiA9L:bb6un1j) ([raw file](https://github.com/zanata/zanata-platform/blob/4c3b9a357aece92d15c408fc4fac3eabc8c10c74/server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java#L125)):*
<details><summary><i>Previously, seanf (Sean Flanigan) wrote…</i></summary><blockquote>

That still looks like an unclosed FileInputStream.

Try something like this:

    InputStream stream = new ByteArrayInputStream(plaintextContent.getBytes(StandardCharsets.UTF_8));
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io/reviews/zanata/zanata-platform/858)*

**Comment by seanf** [comment]
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java, line 51 at r3](https://reviewable.io/reviews/zanata/zanata-platform/858#-LDoPmc02-yaNjoeSuX5:-LDoP7CS8pcZb1EvEqBe:b-sjg71k) ([raw file](https://github.com/zanata/zanata-platform/blob/7b0600819c5e4829554e901a575b4714eb510e01/server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java#L51)):*
> ```Java
> import java.io.ByteArrayInputStream;
> import java.io.File;
> import java.io.FileInputStream;
> ```

You shouldn't need to import File or FileInputStream in this test if you use ByteArrayInputStream and NullInputStream.

---

*[server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java, line 98 at r3](https://reviewable.io/reviews/zanata/zanata-platform/858#-LDoPIhj9aRfwAinbhoI:-LDoPIhj9aRfwAinbhoJ:b-jke92z) ([raw file](https://github.com/zanata/zanata-platform/blob/7b0600819c5e4829554e901a575b4714eb510e01/server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java#L98)):*
> ```Java
>                 "msgstr[1] \"%d aoeuaouao\"\n" +
>                 "msgstr[2] \"\"";
>         File tempFile = File.createTempFile("test", ".po");
> ```

These tests are still creating temp files, but only the names are actually needed. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/zanata/zanata-platform/858)*

**Comment by djansen-redhat** [comment]
@seanf updated.

**Comment by seanf** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, all discussions resolved.

---

*[server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java, line 96 at r4](https://reviewable.io/reviews/zanata/zanata-platform/858#-LDqQy5l1KqfD27immna:-LDqQy5m67bKr15wJbx-:b-rj0pdb) ([raw file](https://github.com/zanata/zanata-platform/blob/58fbe9cd48e481d4eea3448eac7ad556a38b008b/server/services/src/test/java/org/zanata/service/impl/TranslationFileServiceImplTest.java#L96)):*
> ```Java
>                 "msgstr[1] \"%d aoeuaouao\"\n" +
>                 "msgstr[2] \"\"";
>         File tempFile = File.createTempFile("test", ".po");
> ```

You missed a spot. But to save time, I'll get rid of this one and let Jenkins get started.

---


*Comments from [Reviewable](https://reviewable.io/reviews/zanata/zanata-platform/858#-:-LDqSAPQ-nb6h4ihjr35:bnfp4nl)*

**Comment by djansen-redhat** [comment]
Oh for ---- sake. Thanks @seanf

**Pull request merged**